addons/website_blog/controllers/main.py

```diff
@@ -259,16 +248,18 @@ def blog_post(self, blog, blog_post, tag_id=None, page=1, enable_editor=None, **
             'next_post': next_post,
             'date': date_begin,
             'blog_url': blog_url,
-            'pager': pager,
-            'comments': comments,
         }
         response = request.render("website_blog.blog_post_complete", values)
 
-        request.session[request.session.sid] = request.session.get(request.session.sid, [])
-        if not (blog_post.id in request.session[request.session.sid]):
-            request.session[request.session.sid].append(blog_post.id)
+        if blog_post.id not in request.session.get('posts_viewed', []):
+            if not request.session.get('posts_viewed'):
+                request.session['posts_viewed'] = []
+
+            request.session['posts_viewed'].append(blog_post.id)
+            request.session.modified = True
+
             # Increase counter
-            blog_post.sudo().write({
+            blog_post.sudo()._write({
```

Comment:
This often generates concurrent update error on hot articles.
Maybe go for smth like `res.users.log` or `SELECT FOR UPDATE` at least?

Reply:
You are right, we should implement a more resilient way to that. An extra model might be a bit overkill but it would be the proper way, in the mean time we could mitigate with a SELECT FOR UPDATE as you suggests.

Reply:
`SELECT FOR UPDATE` is no different from `UPDATE` in this case, you probably meant `SELECT FOR UPDATE NOWAIT` and then catch the exception and skip updating `visits`.

A simpler way to do this without needing exception handling is to use PG 9.5's SKIP LOCKED clause, which means we only need 1 query and no try/except:

```sql
UPDATE blog_post SET visits = visits + 1
 WHERE id IN (SELECT id from blog_post WHERE id = %s FOR UPDATE SKIP LOCKED);
```

This will update 1 or 0 records depending if any other transaction is writing on the same post at the same time. Obviously this means we will not be counting all views, especially during high activity on a post, but we'll avoid some delays and retries. It's a trade-off I can live with.

Reply:
Yeah, I meant `NOWAIT` as it is already used in Odoo.
I prefer `SKIP LOCKED` for concurrent queues. Your example requires invalidating cache after sql query so it will have negative impact too.

Reply:
Well, both cases require flushing counter value.